pkg/engine/executor/evaluator.go

```diff
@@ -0,0 +1,111 @@
+package executor
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"iter"
+
+	"github.com/apache/arrow-go/v18/arrow/memory"
+	"github.com/grafana/loki/v3/pkg/engine/planner/physical"
+)
+
+// TODO(rfratto): document constraints on how to properly write process
+// methods; particularly around guaranteeing that no Record leaks:
+//
+// * Each process method iterator must yield an error OR generate and yield at
+//   most one Record. Iterators may not generate multiple Records in advance.
+//
+// * Each iterator consumer must Release() the Record after it is done with it.
+//
+// It is safe for iterators to stop iterating early if yield returns false, as
+// future Records won't have been generated yet.
+
+type evaluator struct {
+	opts Options
+	plan *physical.Plan
+	mem  memory.Allocator
+}
+
+func newEvaluator(opts Options, plan *physical.Plan) *evaluator {
+	return &evaluator{opts: opts, plan: plan}
+}
+
+func (e *evaluator) processNode(ctx context.Context, n physical.Node) iter.Seq[Result] {
+	children := e.plan.Children(n)
+
+	iters := make([]iter.Seq[Result], len(children))
+	for i, child := range children {
+		iters[i] = e.processNode(ctx, child)
+	}
```

Comment:
From my experience with the old engine this is a little tricky because it's not clear when nodes are evaluated and results are created. Is it lazy or not? E.g. the underlying method could aggregate all results into a slice and return that or it could create a result on each `yield` call.